### PR TITLE
Fix inaccurate round while being fast

### LINKED ISSUE #7306: Floating point round function gives some inaccurate results
This bug is branched from rust-lang/rust#55107.

The heart of the implementation of `round()` in src/library.js is:

```javascript
   return d >= +0 ? +Math_floor(d + +0.5) : +Math_ceil(d - +0.5);
```

This produces incorrect values for 0.49999999999999994 (correct 0.0, actual 1.0) and 4503599627370497.0 (correct same as input, actual 4503599627370498.0) among others. The latter is disturbing because it means the function does not reliably preserve integers.

I worked out a fix, which is basically this in the positive branch:

```javascript
        Math_floor((d + (0.25 - 0.5 * F64_EPSILON)) + (0.25 + 0.5 * F64_EPSILON))
```
where F64_EPSILON is is the difference between 1.0 and the next largest representable number, or 2.2204460492503131e-16. At least this gives correct results on my x86_64; I'm not sure to what extent correctness is guaranteed across all target platforms.

Also, the ternary might profitably be replaced by (this is C syntax, because I know the libm names of things but not necessarily the relevant library functions/intrinsics in Emscripten):

```c
    copysign(floor(((abs(d) + (0.25 - 0.5 * F64_EPSILON)) + (0.25 + 0.5 * F64_EPSILON)), d)
```

This will probably be faster in wasm because there is a copysign intrinsic. But if the intrinsic doesn't get compiled properly it's probably an overall lose.

I'm willing to work on this, but wanted to give people a heads-up, and maybe could get some guidance on the copysign question.

### PR DESCRIPTION
**Pull request by raphlinus**
This patch applies a clever rounding approach to fix subtle correctness
bugs in the implementation of float rounding. It uses copysign to avoid
branching, and with the hope of using llvm intrinsics where available.

Work in progress, tests are missing.

Fixes #7306

### CODE REVIEW

File: src/library.js [hunk 1]
```diff
@@ -1566,30 +1566,34 @@ LibraryManager.library = {
 
   round__asm: true,
   round__sig: 'dd',
+  round__deps: ['llvm_copysign_f64'],
   round: function(d) {
     d = +d;
-    return d >= +0 ? +Math_floor(d + +0.5) : +Math_ceil(d - +0.5);
+    return +Math_trunc(d + +_llvm_copysign_f64(+0.49999999999999994, d));
```
Comment (kripken) [inline src/library.js]: Note that this is handwritten JS code, and in particular, `llvm_copysign_*` is present here as a fallback, implementing it in JS when it can't be done otherwise. It's a slow code path, see https://github.com/kripken/emscripten/blob/incoming/src/library.js#L1559

To actually use copysign, this would need to be written in C, perhaps in `./system/lib/libc/extras.c`, and it could use a copysign intrinsic there.

Comment (kripken) [reply src/library.js]: I think the CI failures are because this function is marked as asm.js, and I don't think `trunc` is a math intrinsic there.

Comment (raphlinus) [reply src/library.js]: Yeah, between these two things and f32, it sounds like writing it in C is the better choice.

I could do it here with a ternary and floor, but it would be slightly inferior wasm, with a branch. I'll have a try at the C version.

**Comment by juj** [comment]
#7306 refers to all sorts of corner cases. It would be good to add tests to the test suite to contain the different corner cases, otherwise the change in this PR is a bit intractable.
